blockchain/api.go

```diff
@@ -1352,29 +1369,165 @@ func NewSimpleGatekeeper(address common.Address, opts *chainOpts) (SimpleGatekee
 	opts := api.opts.getTxOpts(ctx, key, api.opts.gasLimit)
-	return api.contract.Payout(opts, to, value, txNumber)
+
+	tx, err := txRetryWrapper(func() (*types.Transaction, error) {
+		return api.contract.Payout(opts, to, value, txNumber)
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	rec, err := WaitTransactionReceipt(ctx, api.client, api.opts.blockConfirmations, api.opts.logParsePeriod, tx)
+	if err != nil {
+		return 0, err
+	}
+
+	for _, l := range rec.Logs {
+		switch l.Topics[0] {
+		case CommitTopic:
+			return Committed, nil
+		case PayoutTopic:
+			return Payouted, nil
+		default:
+			continue
+		}
+	}
+	return UNKNOWN, fmt.Errorf("transaction executed successfuly, but payout's topic not found")
 }
 
 func (api *BasicSimpleGatekeeper) Kill(ctx context.Context, key *ecdsa.PrivateKey) (*types.Transaction, error) {
 	opts := api.opts.getTxOpts(ctx, key, api.opts.gasLimit)
 	return api.contract.Kill(opts)
 }
+
+func (api *BasicSimpleGatekeeper) GetPayinTransactions(ctx context.Context) (map[string]*GateTx, error) {
+	logs, err := api.client.FilterLogs(context.Background(), ethereum.FilterQuery{
+		Topics:    [][]common.Hash{{PayinTopic}},
+		Addresses: []common.Address{api.address},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return api.transformTransactionsLogsToMap(logs, PayinTopic), nil
+}
+
+func (api *BasicSimpleGatekeeper) GetGateTransactionTime(ctx context.Context, tx *GateTx) (uint64, error) {
+	block, err := api.client.BlockByNumber(ctx, big.NewInt(0).SetUint64(tx.BlockNumber))
+	if err != nil {
+		return 0, err
+	}
+	if block.Time().IsUint64() {
+		return block.Time().Uint64(), nil
+	} else {
+		return 0, errors.New("block time overflows uint64")
+	}
+}
+
+func (api *BasicSimpleGatekeeper) GetPayoutTransactions(ctx context.Context) (map[string]*GateTx, error) {
+	logs, err := api.client.FilterLogs(context.Background(), ethereum.FilterQuery{
+		Topics:    [][]common.Hash{{PayoutTopic}},
+		Addresses: []common.Address{api.address},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return api.transformTransactionsLogsToMap(logs, PayoutTopic), nil
+}
+
+func (api *BasicSimpleGatekeeper) transformTransactionsLogsToMap(logs []types.Log, topic common.Hash) map[string]*GateTx {
+	transactions := make(map[string]*GateTx)
+	for _, l := range logs {
+		if len(l.Topics) < 4 {
+			continue
+		}
+		if l.Topics[0].String() == topic.String() {
+			transactions[l.Topics[2].Big().String()] = &GateTx{
+				From:        common.HexToAddress(l.Topics[1].Hex()),
+				Number:      l.Topics[2].Big(),
+				Value:       l.Topics[3].Big(),
+				BlockNumber: l.BlockNumber,
+			}
+		}
+	}
+	return transactions
+}
+
+func (api *BasicSimpleGatekeeper) GetTransactionState(ctx context.Context, from common.Address, value *big.Int, txNumber *big.Int) (*GateTxState, error) {
+	id := api.generateTransactionID(from, value, txNumber)
+	t, err := api.contract.Paid(&bind.CallOpts{Pending: true, Context: ctx}, id)
```

Comment:
pending: true? WTF?

Reply:
yeap, need to fix it.
talk over it yesterday:
it is possible to be a feature - to learn that someone already pays the transaction and does not try to pay it again to be free of race.

Reply:
being fucked out with the unfixed transaction state - exactly not a feature, imo.